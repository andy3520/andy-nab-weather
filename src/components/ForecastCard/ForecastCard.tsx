import React from 'react';
import { WEATHER_STATES } from '../../constants/weather';
import { DisplayWeatherDetail, IWeatherLoc } from '../../types';
import { getRoundNum } from '../../utils/getRoundNum';
import { getWeekDay } from '../../utils/getWeekDay';

export interface IForecastCardProps {
  currentWeatherIndex: number;
  weather: IWeatherLoc;
}

const ForecastCard: React.FC<IForecastCardProps> = ({
  currentWeatherIndex,
  weather,
}) => {
  const currentWeather: DisplayWeatherDetail =
    weather?.consolidated_weather?.[currentWeatherIndex];

  if (!currentWeather) return null;

  const { title: location } = weather;
  const {
    applicable_date,
    max_temp,
    min_temp,
    the_temp,
    weather_state_abbr,
  } = currentWeather;
  const weatherState = WEATHER_STATES?.[weather_state_abbr];

  return (
    <div className="px-4 py-5 rounded-lg shadow bg-gradient-to-br sm:bg-gradient-to-r from-indigo-600 to-blue-500 sm:px-6">
      <div className="flex flex-wrap items-center justify-center sm:justify-between">
        <div className="flex-1 text-white md:pr-6">
          <h2 className="text-lg font-semibold sm:text-xl">
            {location} Weather
          </h2>
          <p className="text-base text-gray-100 sm:text-lg">
            {getWeekDay(applicable_date)} - {applicable_date}
          </p>
          <p className="font-bold uppercase text-7xl sm:text-9xl">
            {getRoundNum(the_temp)}°
          </p>
          <p className="text-xl font-semibold text-gray-100 sm:text-2xl">
            {getRoundNum(min_temp)}° &#8644; {getRoundNum(max_temp)}°
          </p>
        </div>
        <img
          src={`/static/images/weather/${weather_state_abbr}.svg`}
          title={weatherState}
          alt={`${weatherState} svg icon`}
          className="w-20 h-20 mt-6 sm:w-40 sm:h-40 sm:mt-0 animate-pulse"
        />
      </div>
    </div>
  );
};

export default ForecastCard;
