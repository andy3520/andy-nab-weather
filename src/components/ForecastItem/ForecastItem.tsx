import React from 'react';
import { WEATHER_STATES } from '../../constants/weather';
import { DisplayWeatherDetail, IConsolidatedWeather } from '../../types';
import { getRoundNum } from '../../utils/getRoundNum';
import { getWeekDay } from '../../utils/getWeekDay';

export interface IForecastItemProps {
  onChangeCurrentWeatherIndex: () => void;
  consolidatedWeather: IConsolidatedWeather | undefined;
}

export enum EForecastItemTestId {
  BUTTON = 'ForecastItem_BUTTON',
  DATE = 'ForecastItem_DATE',
  WEEK_DAY = 'ForecastItem_WEEK_DAY',
  THE_TEMP = 'ForecastItem_THE_TEMP',
  MIN_TEMP = 'ForecastItem_MIN_TEMP',
  MAX_TEMP = 'ForecastItem_MAX_TEMP',
  IMAGE = 'ForecastItem_IMAGE',
}

const ForecastItem: React.FC<IForecastItemProps> = ({
  onChangeCurrentWeatherIndex,
  consolidatedWeather,
}) => {
  if (!consolidatedWeather) return null;

  const {
    applicable_date,
    max_temp,
    min_temp,
    the_temp,
    weather_state_abbr,
  }: DisplayWeatherDetail = consolidatedWeather;

  const weatherState = WEATHER_STATES?.[weather_state_abbr];
  const srcWeatherIcon = `/static/images/weather/${weather_state_abbr}.svg`;

  return (
    <button
      data-testid={EForecastItemTestId.BUTTON}
      onClick={onChangeCurrentWeatherIndex}
      type="button"
      className="flex-1 p-4 rounded-lg shadow bg-gradient-to-r from-indigo-500 to-blue-400 hover:from-indigo-400 hover:to-blue-300 focus:from-indigo-400 focus: focus:to-blue-300 focus:ring-4 focus:ring-indigo-200 focus:outline-none"
    >
      <div className="flex flex-col items-center justify-center text-white">
        <p
          className="font-semibold text-gray-100"
          data-testid={EForecastItemTestId.WEEK_DAY}
        >
          {getWeekDay(applicable_date)}
        </p>
        <p
          className="text-xs text-gray-200"
          data-testid={EForecastItemTestId.DATE}
        >
          {applicable_date}
        </p>
        <div className="flex items-center justify-between">
          <p
            className="text-3xl font-bold"
            data-testid={EForecastItemTestId.THE_TEMP}
          >
            {getRoundNum(the_temp)}°
          </p>
          <img
            data-testid={EForecastItemTestId.IMAGE}
            title={weatherState}
            src={srcWeatherIcon}
            alt={`${weatherState} icon`}
            className="ml-2 w-7 h-7"
          />
        </div>
        <p className="text-sm text-gray-100">
          <span data-testid={EForecastItemTestId.MIN_TEMP}>
            {getRoundNum(min_temp)}
          </span>
          ° &#8644;{' '}
          <span data-testid={EForecastItemTestId.MAX_TEMP}>
            {getRoundNum(max_temp)}
          </span>
          °
        </p>
      </div>
    </button>
  );
};

export default ForecastItem;
