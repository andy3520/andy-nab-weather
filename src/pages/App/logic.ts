import React from 'react';
import { EApiMessages } from '../../constants/api';
import { fetchIpLoc, fetchLoc, fetchLocWeather } from '../../services/api';
import { IIPInfo, IWeatherLoc, IWeatherLocSearch } from '../../types';

export interface IFetchHandlerParams {
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
  setErrorMsg: React.Dispatch<React.SetStateAction<string | undefined>>;
  callback?: CallableFunction;
}
export const fetchHandler = async ({
  setLoading,
  setErrorMsg,
  callback,
}: IFetchHandlerParams): Promise<void> => {
  setErrorMsg('');
  setLoading(true);

  try {
    if (callback && typeof callback === 'function') {
      await callback();
    }
  } catch (error) {
    setErrorMsg(error?.message);
  } finally {
    setLoading(false);
  }
};

export type DetectLocationParams = Omit<IFetchHandlerParams, 'callback'> & {
  setIPLoc: React.Dispatch<React.SetStateAction<IIPInfo | undefined>>;
};
export const detectLoc = async ({
  setLoading,
  setErrorMsg,
  setIPLoc,
}: DetectLocationParams): Promise<void> => {
  await fetchHandler({
    setLoading,
    setErrorMsg,
    callback: async () => {
      setIPLoc(await fetchIpLoc());
    },
  });
};

export type LoadDetectedLocationParams = Omit<
  IFetchHandlerParams,
  'callback'
> & {
  setCurrentLoc: React.Dispatch<
    React.SetStateAction<IWeatherLocSearch | undefined>
  >;
  setNotifyMsg: React.Dispatch<React.SetStateAction<string | undefined>>;
  city: string;
  ip: string;
};
export const loadDetectedLocation = async ({
  setLoading,
  setErrorMsg,
  setCurrentLoc,
  setNotifyMsg,
  city,
  ip,
}: LoadDetectedLocationParams): Promise<void> => {
  await fetchHandler({
    setLoading,
    setErrorMsg,
    callback: async () => {
      const locations = await fetchLoc({ queryName: city });

      if (!locations?.length)
        setNotifyMsg(
          `Not found forecast for detected location: ${city} - by IP: ${ip}.\n Please try to use search.`
        );

      setCurrentLoc(locations[0]);
    },
  });
};

export type FetchCurrentLocWeatherParams = Omit<
  IFetchHandlerParams,
  'callback'
> & {
  setWeather: React.Dispatch<React.SetStateAction<IWeatherLoc | undefined>>;
  woeid: number;
};
export const fetchCurrentLocWeather = async ({
  setLoading,
  setErrorMsg,
  setWeather,
  woeid,
}: FetchCurrentLocWeatherParams): Promise<void> => {
  await fetchHandler({
    setErrorMsg,
    setLoading,
    callback: async () => {
      const weather = await fetchLocWeather({ woeid });
      setWeather(weather);
    },
  });
};

export type OnSearchParams = Omit<IFetchHandlerParams, 'callback'> & {
  setWeather: React.Dispatch<React.SetStateAction<IWeatherLoc | undefined>>;
  setNotifyMsg: React.Dispatch<React.SetStateAction<string | undefined>>;
};
export const onSearch = ({
  setLoading,
  setErrorMsg,
  setWeather,
  setNotifyMsg,
}: OnSearchParams) => async (
  event: React.FormEvent<HTMLInputElement>
): Promise<void> => {
  setNotifyMsg('');
  await fetchHandler({
    setLoading,
    setErrorMsg,
    callback: async () => {
      const queryName = (event.target as HTMLInputElement).value;

      if (!queryName) return;
      setWeather(undefined);

      const encodedQuery = encodeURIComponent(queryName);
      const locations = await fetchLoc({ queryName: encodedQuery });
      if (!locations?.length) throw new Error(EApiMessages.NOT_FOUND);

      const firstLocWoeId = locations[0]?.woeid;
      if (!firstLocWoeId) throw new Error(EApiMessages.NOT_FOUND);

      const weather = await fetchLocWeather({ woeid: firstLocWoeId });
      setWeather(weather);
    },
  });
};
