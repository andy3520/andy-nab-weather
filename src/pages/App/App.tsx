import React from 'react';
import './App.scss';
import { ReactComponent as GithubIcon } from './github.svg';
import { ReactComponent as TwitterIcon } from './twitter.svg';
import { ReactComponent as LoaderIcon } from './loader.svg';
import ForecastCard from '../../components/ForecastCard';
import ForecastList from '../../components/ForecastList';
import LocationSearch from '../../components/LocationSearch';
import Notify from '../../components/Notify';
import { IIPInfo, IWeatherLoc, IWeatherLocSearch } from '../../types';
import { EApiMessages } from '../../constants/api';
import {
  detectLoc,
  fetchCurrentLocWeather,
  loadDetectedLocation,
  onSearch,
} from './app.func';

export enum EAppTestId {
  LOADER = 'App_LOADER',
}

const App: React.FC = () => {
  const [ipLoc, setIPLoc] = React.useState<IIPInfo>();
  const [currentLoc, setCurrentLoc] = React.useState<IWeatherLocSearch>();
  const [weather, setWeather] = React.useState<IWeatherLoc>();
  const [currentWeatherIndex, setCurrentWeatherIndex] = React.useState<number>(
    0
  );

  const [isLoading, setLoading] = React.useState<boolean>(true);
  const [errorMsg, setErrorMsg] = React.useState<string>();
  const [notifyType, setNotifyType] = React.useState<'info' | 'error'>();
  const [notifyMsg, setNotifyMsg] = React.useState<string>();

  React.useEffect(() => {
    detectLoc({
      setLoading,
      setErrorMsg,
      setIPLoc,
    });
  }, []);

  React.useEffect(() => {
    if (!ipLoc?.city || !ipLoc?.ip) {
      setNotifyType('info');
      setNotifyMsg(EApiMessages.CANNOT_DETECT_LOCATION);
      return;
    }

    setNotifyType('info');
    setNotifyMsg(`Detected location: ${ipLoc.city} from ip ${ipLoc.ip}`);
    loadDetectedLocation({
      setLoading,
      setErrorMsg,
      setCurrentLoc,
      setNotifyMsg,
      city: ipLoc.city,
      ip: ipLoc.ip,
    });
  }, [ipLoc]);

  React.useEffect(() => {
    if (currentLoc?.woeid) {
      fetchCurrentLocWeather({
        setLoading,
        setErrorMsg,
        setWeather,
        woeid: currentLoc?.woeid,
      });

      setCurrentWeatherIndex(0);
    }
  }, [currentLoc]);

  React.useEffect(() => {
    setNotifyType('error');
    setNotifyMsg(errorMsg);
  }, [errorMsg]);

  return (
    <div className="bg-gray-500">
      <div className="container">
        <div className="flex items-center justify-center min-h-screen p-4">
          <div className="w-full max-w-4xl overflow-hidden bg-white divide-y divide-gray-200 rounded-lg shadow">
            <header className="flex items-center justify-between px-4 py-5 sm:px-6">
              <h1 className="font-mono text-xl font-bold tracking-wider uppercase">
                Andy&apos;s Forecast
              </h1>
              <LocationSearch
                onSearch={onSearch({
                  setLoading,
                  setErrorMsg,
                  setWeather,
                  setNotifyMsg,
                })}
              />
            </header>
            <div className="px-4 py-5 sm:p-6">
              {notifyMsg && (
                <div className="mb-4">
                  <Notify type={notifyType} msg={notifyMsg} />
                </div>
              )}

              {isLoading && (
                <LoaderIcon
                  data-testid={EAppTestId.LOADER}
                  className="w-20 h-20 text-indigo-600 animate-spin"
                />
              )}

              {weather && (
                <ForecastCard
                  currentWeatherIndex={currentWeatherIndex}
                  weather={weather}
                />
              )}

              {weather && (
                <div className="mt-4 sm:mt-6">
                  <ForecastList
                    weather={weather}
                    setCurrentWeatherIndex={setCurrentWeatherIndex}
                  />
                </div>
              )}
            </div>
            <footer className="px-4 py-4 sm:px-6">
              <div className="flex items-center justify-between">
                <div className="font-mono text-sm font-semibold">
                  NAB Assignment by Andy Nguyen with ❤️
                </div>
                <div className="flex">
                  <a
                    href="https://twitter.com/annguyenhieuduc"
                    className="mr-4 text-gray-400 hover:text-gray-500"
                    target="_blank"
                    rel="noreferrer"
                  >
                    <span className="sr-only">Twitter</span>
                    <TwitterIcon className="w-6 h-6" />
                  </a>

                  <a
                    href="https://github.com/andy3520/andy-nab-weather"
                    target="_blank"
                    rel="noreferrer"
                    className="text-gray-400 hover:text-gray-500"
                  >
                    <span className="sr-only">GitHub</span>
                    <GithubIcon className="w-6 h-6" />
                  </a>
                </div>
              </div>
            </footer>
          </div>
        </div>
      </div>
    </div>
  );
};

export default App;
