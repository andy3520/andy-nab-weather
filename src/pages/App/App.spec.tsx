import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import {
  render,
  fireEvent,
  waitFor,
  RenderResult,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import _fetch from 'isomorphic-fetch';
import _fetchMock from 'fetch-mock';
import {
  ipLocSample,
  locSample,
  locWeatherSample,
} from '../../constants/testSample';
import App, { EAppTestId } from './App';
import { EForecastCardTestId } from '../../components/ForecastCard/ForecastCard';
import { EForecastItemTestId } from '../../components/ForecastItem/ForecastItem';

const fetchMock = (_fetch as unknown) as typeof _fetchMock;

const mockNormalApi = () => {
  const ipLocMatcher = /check/;
  const locMather = /api\/location\/search\/\?query=/;
  const weathMatcher = /api\/location\/\d+/;
  const ipLocJSON = JSON.stringify(ipLocSample);
  const locSampleJSON = JSON.stringify(locSample);
  const locWeatherSampleJSON = JSON.stringify(locWeatherSample);
  fetchMock.mock(ipLocMatcher, {
    status: 200,
    body: ipLocJSON,
  });
  fetchMock.mock(locMather, {
    status: 200,
    body: locSampleJSON,
  });
  fetchMock.mock(weathMatcher, {
    status: 200,
    body: locWeatherSampleJSON,
  });
};

describe('App', () => {
  afterEach(() => {
    fetchMock.reset();
  });
  test('should match snapshot', async () => {
    mockNormalApi();
    const { asFragment, getByTestId } = render(<App />);
    await waitForElementToBeRemoved(() => getByTestId(EAppTestId.LOADER));
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render data correctly', async () => {
    mockNormalApi();
    const { getByTestId, getAllByTestId } = render(<App />);
    await waitForElementToBeRemoved(() => getByTestId(EAppTestId.LOADER));

    const TITLE = getByTestId(EForecastCardTestId.TITLE);
    const ITEMS = getAllByTestId(EForecastItemTestId.BUTTON);

    const expectedTitle = `${locWeatherSample.title} Weather`;

    expect(ITEMS).toHaveLength(locWeatherSample.consolidated_weather.length);
    expect(TITLE).toHaveTextContent(expectedTitle);
  });
});
