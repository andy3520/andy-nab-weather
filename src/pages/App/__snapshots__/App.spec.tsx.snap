// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-gray-500"
  >
    <div
      class="container"
    >
      <div
        class="flex items-center justify-center min-h-screen p-4"
      >
        <div
          class="w-full max-w-4xl overflow-hidden bg-white divide-y divide-gray-200 rounded-lg shadow"
        >
          <header
            class="flex items-center justify-between px-4 py-5 sm:px-6"
          >
            <h1
              class="font-mono text-xl font-bold tracking-wider uppercase"
            >
              Andy's Forecast
            </h1>
            <div
              class="w-full max-w-lg lg:max-w-xs"
            >
              <label
                class="sr-only"
                for="search"
              >
                Search
              </label>
              <div
                class="relative"
              >
                <div
                  class="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none"
                >
                  <svg
                    class="w-5 h-5 text-gray-400"
                    data-testid="LocationSearch_ICON"
                  >
                    search.svg
                  </svg>
                </div>
                <input
                  class="block w-full py-2 pl-10 pr-3 leading-5 placeholder-gray-500 bg-white border border-gray-300 rounded-md focus:outline-none focus:placeholder-gray-400 focus:ring-1 focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
                  data-testid="LocationSearch_INPUT"
                  id="search"
                  name="search"
                  placeholder="City name"
                  type="search"
                />
              </div>
            </div>
          </header>
          <div
            class="px-4 py-5 sm:p-6"
          >
            <div
              class="mb-4"
            >
              <div
                class="bg-indigo-200 text-indigo-600 shadow px-4 py-5 sm:px-6 rounded-lg"
              >
                Detected location: Ho Chi Minh City from ip 113.185.78.133
              </div>
            </div>
            <div
              class="px-4 py-5 rounded-lg shadow bg-gradient-to-br sm:bg-gradient-to-r from-indigo-600 to-blue-500 sm:px-6"
            >
              <div
                class="flex flex-wrap items-center justify-center sm:justify-between"
              >
                <div
                  class="flex-1 text-white md:pr-6"
                >
                  <h2
                    class="text-lg font-semibold sm:text-xl"
                    data-testid="ForecastCard_TITLE"
                  >
                    Ho Chi Minh City Weather
                  </h2>
                  <p
                    class="text-base text-gray-100 sm:text-lg"
                  >
                    <span
                      data-testid="ForecastCard_WEEK_DAY"
                    >
                      Friday
                    </span>
                     - 
                    <span
                      data-testid="ForecastCard_DATE"
                    >
                      2021-01-01
                    </span>
                  </p>
                  <p
                    class="font-bold uppercase text-7xl sm:text-9xl"
                  >
                    <span
                      data-testid="ForecastCard_THE_TEMP"
                    >
                      28
                    </span>
                    °
                  </p>
                  <p
                    class="text-xl font-semibold text-gray-100 sm:text-2xl"
                  >
                    <span
                      data-testid="ForecastCard_MIN_TEMP"
                    >
                      21
                    </span>
                    ° ⇄ 
                    <span
                      data-testid="ForecastCard_MAX_TEMP"
                    >
                      29
                    </span>
                    °
                  </p>
                </div>
                <img
                  alt="Heavy Cloud icon"
                  class="w-20 h-20 mt-6 sm:w-40 sm:h-40 sm:mt-0 animate-pulse"
                  data-testid="ForecastCard_IMAGE"
                  src="/static/images/weather/hc.svg"
                  title="Heavy Cloud"
                />
              </div>
            </div>
            <div
              class="mt-6"
            >
              <div
                class="flex flex-wrap items-center justify-between -m-2"
              >
                <div
                  class="p-2"
                  data-testid="ForecastList_ITEM"
                >
                  <button
                    class="flex-1 p-4 rounded-lg shadow bg-gradient-to-r from-indigo-500 to-blue-400 hover:from-indigo-400 hover:to-blue-300 focus:from-indigo-400 focus: focus:to-blue-300 focus:ring-4 focus:ring-indigo-200 focus:outline-none"
                    data-testid="ForecastItem_BUTTON"
                    type="button"
                  >
                    <div
                      class="flex flex-col items-center justify-center text-white"
                    >
                      <p
                        class="font-semibold text-gray-100"
                        data-testid="ForecastItem_WEEK_DAY"
                      >
                        Friday
                      </p>
                      <p
                        class="text-xs text-gray-200"
                        data-testid="ForecastItem_DATE"
                      >
                        2021-01-01
                      </p>
                      <div
                        class="flex items-center justify-between"
                      >
                        <p
                          class="text-3xl font-bold"
                          data-testid="ForecastItem_THE_TEMP"
                        >
                          28°
                        </p>
                        <img
                          alt="Heavy Cloud icon"
                          class="ml-2 w-7 h-7"
                          data-testid="ForecastItem_IMAGE"
                          src="/static/images/weather/hc.svg"
                          title="Heavy Cloud"
                        />
                      </div>
                      <p
                        class="text-sm text-gray-100"
                      >
                        <span
                          data-testid="ForecastItem_MIN_TEMP"
                        >
                          21
                        </span>
                        ° ⇄ 
                        <span
                          data-testid="ForecastItem_MAX_TEMP"
                        >
                          29
                        </span>
                        °
                      </p>
                    </div>
                  </button>
                </div>
                <div
                  class="p-2"
                  data-testid="ForecastList_ITEM"
                >
                  <button
                    class="flex-1 p-4 rounded-lg shadow bg-gradient-to-r from-indigo-500 to-blue-400 hover:from-indigo-400 hover:to-blue-300 focus:from-indigo-400 focus: focus:to-blue-300 focus:ring-4 focus:ring-indigo-200 focus:outline-none"
                    data-testid="ForecastItem_BUTTON"
                    type="button"
                  >
                    <div
                      class="flex flex-col items-center justify-center text-white"
                    >
                      <p
                        class="font-semibold text-gray-100"
                        data-testid="ForecastItem_WEEK_DAY"
                      >
                        Staturday
                      </p>
                      <p
                        class="text-xs text-gray-200"
                        data-testid="ForecastItem_DATE"
                      >
                        2021-01-02
                      </p>
                      <div
                        class="flex items-center justify-between"
                      >
                        <p
                          class="text-3xl font-bold"
                          data-testid="ForecastItem_THE_TEMP"
                        >
                          30°
                        </p>
                        <img
                          alt="Heavy Cloud icon"
                          class="ml-2 w-7 h-7"
                          data-testid="ForecastItem_IMAGE"
                          src="/static/images/weather/hc.svg"
                          title="Heavy Cloud"
                        />
                      </div>
                      <p
                        class="text-sm text-gray-100"
                      >
                        <span
                          data-testid="ForecastItem_MIN_TEMP"
                        >
                          22
                        </span>
                        ° ⇄ 
                        <span
                          data-testid="ForecastItem_MAX_TEMP"
                        >
                          30
                        </span>
                        °
                      </p>
                    </div>
                  </button>
                </div>
                <div
                  class="p-2"
                  data-testid="ForecastList_ITEM"
                >
                  <button
                    class="flex-1 p-4 rounded-lg shadow bg-gradient-to-r from-indigo-500 to-blue-400 hover:from-indigo-400 hover:to-blue-300 focus:from-indigo-400 focus: focus:to-blue-300 focus:ring-4 focus:ring-indigo-200 focus:outline-none"
                    data-testid="ForecastItem_BUTTON"
                    type="button"
                  >
                    <div
                      class="flex flex-col items-center justify-center text-white"
                    >
                      <p
                        class="font-semibold text-gray-100"
                        data-testid="ForecastItem_WEEK_DAY"
                      >
                        Sunday
                      </p>
                      <p
                        class="text-xs text-gray-200"
                        data-testid="ForecastItem_DATE"
                      >
                        2021-01-03
                      </p>
                      <div
                        class="flex items-center justify-between"
                      >
                        <p
                          class="text-3xl font-bold"
                          data-testid="ForecastItem_THE_TEMP"
                        >
                          30°
                        </p>
                        <img
                          alt="Showers icon"
                          class="ml-2 w-7 h-7"
                          data-testid="ForecastItem_IMAGE"
                          src="/static/images/weather/s.svg"
                          title="Showers"
                        />
                      </div>
                      <p
                        class="text-sm text-gray-100"
                      >
                        <span
                          data-testid="ForecastItem_MIN_TEMP"
                        >
                          23
                        </span>
                        ° ⇄ 
                        <span
                          data-testid="ForecastItem_MAX_TEMP"
                        >
                          32
                        </span>
                        °
                      </p>
                    </div>
                  </button>
                </div>
                <div
                  class="p-2"
                  data-testid="ForecastList_ITEM"
                >
                  <button
                    class="flex-1 p-4 rounded-lg shadow bg-gradient-to-r from-indigo-500 to-blue-400 hover:from-indigo-400 hover:to-blue-300 focus:from-indigo-400 focus: focus:to-blue-300 focus:ring-4 focus:ring-indigo-200 focus:outline-none"
                    data-testid="ForecastItem_BUTTON"
                    type="button"
                  >
                    <div
                      class="flex flex-col items-center justify-center text-white"
                    >
                      <p
                        class="font-semibold text-gray-100"
                        data-testid="ForecastItem_WEEK_DAY"
                      >
                        Monday
                      </p>
                      <p
                        class="text-xs text-gray-200"
                        data-testid="ForecastItem_DATE"
                      >
                        2021-01-04
                      </p>
                      <div
                        class="flex items-center justify-between"
                      >
                        <p
                          class="text-3xl font-bold"
                          data-testid="ForecastItem_THE_TEMP"
                        >
                          31°
                        </p>
                        <img
                          alt="Light Rain icon"
                          class="ml-2 w-7 h-7"
                          data-testid="ForecastItem_IMAGE"
                          src="/static/images/weather/lr.svg"
                          title="Light Rain"
                        />
                      </div>
                      <p
                        class="text-sm text-gray-100"
                      >
                        <span
                          data-testid="ForecastItem_MIN_TEMP"
                        >
                          23
                        </span>
                        ° ⇄ 
                        <span
                          data-testid="ForecastItem_MAX_TEMP"
                        >
                          33
                        </span>
                        °
                      </p>
                    </div>
                  </button>
                </div>
                <div
                  class="p-2"
                  data-testid="ForecastList_ITEM"
                >
                  <button
                    class="flex-1 p-4 rounded-lg shadow bg-gradient-to-r from-indigo-500 to-blue-400 hover:from-indigo-400 hover:to-blue-300 focus:from-indigo-400 focus: focus:to-blue-300 focus:ring-4 focus:ring-indigo-200 focus:outline-none"
                    data-testid="ForecastItem_BUTTON"
                    type="button"
                  >
                    <div
                      class="flex flex-col items-center justify-center text-white"
                    >
                      <p
                        class="font-semibold text-gray-100"
                        data-testid="ForecastItem_WEEK_DAY"
                      >
                        Tuesday
                      </p>
                      <p
                        class="text-xs text-gray-200"
                        data-testid="ForecastItem_DATE"
                      >
                        2021-01-05
                      </p>
                      <div
                        class="flex items-center justify-between"
                      >
                        <p
                          class="text-3xl font-bold"
                          data-testid="ForecastItem_THE_TEMP"
                        >
                          28°
                        </p>
                        <img
                          alt="Heavy Rain icon"
                          class="ml-2 w-7 h-7"
                          data-testid="ForecastItem_IMAGE"
                          src="/static/images/weather/hr.svg"
                          title="Heavy Rain"
                        />
                      </div>
                      <p
                        class="text-sm text-gray-100"
                      >
                        <span
                          data-testid="ForecastItem_MIN_TEMP"
                        >
                          24
                        </span>
                        ° ⇄ 
                        <span
                          data-testid="ForecastItem_MAX_TEMP"
                        >
                          31
                        </span>
                        °
                      </p>
                    </div>
                  </button>
                </div>
                <div
                  class="p-2"
                  data-testid="ForecastList_ITEM"
                >
                  <button
                    class="flex-1 p-4 rounded-lg shadow bg-gradient-to-r from-indigo-500 to-blue-400 hover:from-indigo-400 hover:to-blue-300 focus:from-indigo-400 focus: focus:to-blue-300 focus:ring-4 focus:ring-indigo-200 focus:outline-none"
                    data-testid="ForecastItem_BUTTON"
                    type="button"
                  >
                    <div
                      class="flex flex-col items-center justify-center text-white"
                    >
                      <p
                        class="font-semibold text-gray-100"
                        data-testid="ForecastItem_WEEK_DAY"
                      >
                        Wednesday
                      </p>
                      <p
                        class="text-xs text-gray-200"
                        data-testid="ForecastItem_DATE"
                      >
                        2021-01-06
                      </p>
                      <div
                        class="flex items-center justify-between"
                      >
                        <p
                          class="text-3xl font-bold"
                          data-testid="ForecastItem_THE_TEMP"
                        >
                          30°
                        </p>
                        <img
                          alt="Heavy Rain icon"
                          class="ml-2 w-7 h-7"
                          data-testid="ForecastItem_IMAGE"
                          src="/static/images/weather/hr.svg"
                          title="Heavy Rain"
                        />
                      </div>
                      <p
                        class="text-sm text-gray-100"
                      >
                        <span
                          data-testid="ForecastItem_MIN_TEMP"
                        >
                          24
                        </span>
                        ° ⇄ 
                        <span
                          data-testid="ForecastItem_MAX_TEMP"
                        >
                          33
                        </span>
                        °
                      </p>
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <footer
            class="px-4 py-4 sm:px-6"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="font-mono text-sm font-semibold"
              >
                NAB Assignment by Andy Nguyen with ❤️
              </div>
              <div
                class="flex"
              >
                <a
                  class="mr-4 text-gray-400 hover:text-gray-500"
                  href="https://twitter.com/annguyenhieuduc"
                  rel="noreferrer"
                  target="_blank"
                >
                  <span
                    class="sr-only"
                  >
                    Twitter
                  </span>
                  <svg
                    class="w-6 h-6"
                  >
                    twitter.svg
                  </svg>
                </a>
                <a
                  class="text-gray-400 hover:text-gray-500"
                  href="https://github.com/andy3520/andy-nab-weather"
                  rel="noreferrer"
                  target="_blank"
                >
                  <span
                    class="sr-only"
                  >
                    GitHub
                  </span>
                  <svg
                    class="w-6 h-6"
                  >
                    github.svg
                  </svg>
                </a>
              </div>
            </div>
          </footer>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
